packages/x-data-grid-pro/src/hooks/features/serverSideData/useGridDataSource.ts

```diff
@@ -0,0 +1,288 @@
+import * as React from 'react';
+import {
+  useGridApiEventHandler,
+  gridRowsLoadingSelector,
+  useGridApiMethod,
+  GridServerSideGroupNode,
+  GridRowId,
+  useGridSelector,
+} from '@mui/x-data-grid';
+import { gridRowGroupsToFetchSelector, GridStateInitializer } from '@mui/x-data-grid/internals';
+import { GridGetRowsResponse } from '../../../models';
+import { GridPrivateApiPro } from '../../../models/gridApiPro';
+import { DataGridProProcessedProps } from '../../../models/dataGridProProps';
+import {
+  gridGetRowsParamsSelector,
+  gridServerSideDataLoadingSelector,
+  gridServerSideDataErrorsSelector,
+} from './gridServerSideDataSelector';
+import { GridDataSourceApi, GridDataSourcePrivateApi } from './serverSideInterfaces';
+import { runIfServerMode, NestedDataManager, RequestStatus } from './utils';
+
+const INITIAL_STATE = {
+  loading: {},
+  errors: {},
+};
+
+export const dataSourceStateInitializer: GridStateInitializer = (state, _, apiRef) => {
+  apiRef.current.caches.serverSideData = {
+    groupKeys: [],
+  };
+
+  return {
+    ...state,
+    serverSideData: INITIAL_STATE,
+  };
+};
+
+export const useGridDataSource = (
+  privateApiRef: React.MutableRefObject<GridPrivateApiPro>,
+  props: Pick<
+    DataGridProProcessedProps,
+    | 'unstable_dataSource'
+    | 'unstable_onServerSideError'
+    | 'sortingMode'
+    | 'filterMode'
+    | 'paginationMode'
+    | 'treeData'
+  >,
+) => {
+  const nestedDataManager = React.useRef<NestedDataManager>(
+    new NestedDataManager(privateApiRef),
+  ).current;
```

Comment:
This is a good use-case for `useLazyRef`, avoids the `new ...` on each render.

Reply:
My understanding is, `useLazyRef` would be more helpful when the ref initial value might not be available (/possible to compute) on the initial render, but if we can compute it on the initial render (which I assume is the use-case), we could use the `useRef` and the constructor code (or initialization) will [never](https://react.dev/reference/react/useRef#returns) re-run on subsequent renders.

Am I missing something? 🤔

Reply:
Maybe the name isn't optimal, but the hook initializes a ref object once on the first render. It avoids re-allocating a `NestedManager` instance on every render like you're doing here. Example here: https://github.com/mui/material-ui/blob/next/packages/mui-utils/src/useTimeout/useTimeout.ts#L35-L41